src/index.ts

```diff
@@ -306,51 +316,65 @@ QuickSQLite.transactionAsync = (
   };
 
   const commit = () => {
+    if (isFinalized) {
+      throw Error(
+        `Quick SQLite Error: Cannot execute commit on finalized transaction: ${dbName}`
+      );
+    }
     const result = QuickSQLite.execute(dbName, 'COMMIT');
     isFinalized = true;
     return result;
   };
 
   const rollback = () => {
+    if (isFinalized) {
+      throw Error(
+        `Quick SQLite Error: Cannot execute rollback on finalized transaction: ${dbName}`
+      );
+    }
     const result = QuickSQLite.execute(dbName, 'ROLLBACK');
     isFinalized = true;
     return result;
   };
 
-  const tx: PendingTransaction = {
-    start: async () => {
-      try {
-        QuickSQLite.execute(dbName, 'BEGIN TRANSACTION');
-        await callback({
-          commit,
-          execute,
-          executeAsync,
-          rollback,
-        });
-
-        if (!isFinalized) {
-          commit();
-        }
-      } catch (e: any) {
-        if (!isFinalized) {
-          rollback();
-        }
+  return await new Promise((resolve, reject) => {
+    const tx: PendingTransaction = {
+      start: async () => {
+        try {
+          QuickSQLite.execute(dbName, 'BEGIN TRANSACTION');
+          await callback({
+            commit,
+            execute,
+            executeAsync,
+            rollback,
+          });
+
+          if (!isFinalized) {
+            commit();
+          }
 
-        // Do not throw an error, because the transaction is executed with a setImmediate call
-        // This errors are uncatchable
-        // https://stackoverflow.com/questions/51081892/nodejs-asynchronous-exceptions-are-uncatchable
+          resolve();
+        } catch (e) {
+          if (!isFinalized) {
+            try {
+              rollback();
+            } catch (rollbackError) {
+              reject(rollbackError);
+            }
+          }
 
-        // throw e;
-      } finally {
-        locks[dbName].inProgress = false;
-        isFinalized = false;
-        startNextTransaction(dbName);
-      }
-    },
-  };
+          reject(e);
```

Comment:
I'm afraid this will cause the same error as on the code you deleted, code wrapped in setImmediate cannot be catched, have you tried throwing an exception inside somehow?

Reply:
A promise is constructed and returned to the caller of the `transactionAsync` function.
The promise callback immediately runs and puts the tx object on the queue (before returning to the caller).
The queue eventually shifts the first PendingTransaction object from the queue and then calls the `start()` inside the setImmediate callback.
When executing the start method of the PendingTransaction, that function has access to the resolve and reject of the originally returned promise (which is still pending) from the a parent closure.
So when it calls reject or resolve the original promise resolves / rejects.
The setImmediate can thus send errors and rejections to the original caller via the original promise.